csrc/device_lower/pass/circular_buffer.cpp

```diff
@@ -418,6 +418,55 @@ class CloneTmaCircularBufferLoopAndInsertSync
     }
   }
 
+  // If there is a persistent outer-loop, this IfThenElse short-circuits
+  // computation for quantized waves.
+  kir::IfThenElse* createPersistentShortCircuit(
+      ForLoop* outer_loop,
+      ForLoop* inner_loop) {
+    NVF_ERROR(outer_loop != nullptr);
+    NVF_ERROR(inner_loop != nullptr);
+    NVF_ERROR(!outer_loop->iter_domain()->isParallelized());
+    NVF_ERROR(outer_loop->index() != nullptr);
+    NVF_ERROR(outer_loop->stop() != nullptr);
+    NVF_ERROR(inner_loop->index() != nullptr);
+    NVF_ERROR(inner_loop->stop() != nullptr);
+
+    // Only add short-circuit if inner for-loop is grid dimension
+    if (!inner_loop->iter_domain()->isBlockDim()) {
+      return nullptr;
+    }
+
+    const auto& opt =
+        GpuLower::current()->circularBufferInfo().getCircularBufferOptionsFor(
+            outer_loop->iter_domain());
+    // Only add short-circuit for warp specialized circular buffering
+    if (!std::holds_alternative<WarpSpecialized>(opt.type)) {
+      return nullptr;
+    }
+
+    // lhs := (outer_fl->index() * inner_fl->stop() + inner_fl->index())
+    Val* lhs =
+        SimplifyingIrBuilder::mulExpr(outer_loop->index(), inner_loop->stop());
+    lhs = SimplifyingIrBuilder::addExpr(lhs, inner_loop->index());
+
+    // Check that outer_loop matches known invariants for persistent kernel.
+    Expr* def = outer_loop->stop()->definition();
+    NVF_ERROR(def != nullptr);
+    BinaryOp* bop = def->as<BinaryOp>();
+    NVF_ERROR(
+        bop != nullptr && bop->getBinaryOpType() == BinaryOpType::CeilDiv);
+    NVF_ERROR(bop->lhs() != nullptr);
```

Comment:
Can we:
1. Instead of checking the `expr` of extents, check `IterDomain` transformations which is more reliable.
2. If the pattern does not match, return `nullptr` instead of erroring out, so that people can still use circular buffering on other cases without triggering an error.

Reply:
Grid traversal pattern were looking at now:
1. 2D reorder =>     `ceilDiv( (M/first_factor * N/second_factor * first_factor * second_factor), num_sms)`
2. Row-Major reorder => `ceilDiv( (M * column-factor * (N/column-factor)), num_sms )`
3. Column-Major reorder => `ceilDiv( (M/row-factor) * N * row-factor), num_sms )`

Reply:
So, check if for-loop iterDomain is `split(x, ParallelType::BIDx)`?

Reply:
Because your predicate is `outer_fl->index() * inner_fl->stop() + inner_fl->index()`, so I am thinking about something like
```
outer_fl, inner_fl = split(x,  inner_fl->stop())
```
Which in your case is `split(x, ParallelType::BIDx)`.

Reply:
I added the `iterDomain` checks to return `nullptr` before creating the short-circuit `IfThenElse` node. I kept the assertions for `bop` as-is.